### PR TITLE
Add access log formatter support for metadata extension

### PR DESCRIPTION
**Pull request by jbohanon**
# Description

## Code changes
- Adds translation to include the metadata access log formatter extension
- Add testing for metadata and req_without_query formatters
- Add query to test http request util

## Docs changes
- The access logging guide could mention these features, but they're part of upstream envoy access logging docs, so may not be necessary

## Testing steps

1. Install gloo edge with 
```shell
helm install -n gloo-system gloo gloo/gloo --create-namespace
```
2. Add upstream with `glooctl create upstream static --static-hosts jsonplaceholder.typicode.com:80 --name json-upstream`
3. Add virtual service with 
```shell
kubectl apply -f - <<EOF
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: vs
  namespace: gloo-system
spec:
  virtualHost:
    domains:
    - '*'
    routes:
    - matchers:
      - exact: /withMetadata
      options:
        envoyMetadata:
          foo-namespace:
            bar-metadata: I'm on the gloo route!
      routeAction:
        single:
          upstream:
            name: json-upstream
            namespace: gloo-system
    - matchers:
      - exact: /
      routeAction:
        single:
          upstream:
            name: json-upstream
            namespace: gloo-system
EOF
```
4. Patch your gateway resource with 
```yaml
spec:
  options:
    accessLoggingService:
      accessLog:
      - fileSink:
          path: /dev/stdout
          stringFormat: |
            route metadata: "%METADATA(ROUTE:foo-namespace)%"; route metadata drilldown: "%METADATA(ROUTE:foo-namespace:bar-metadata)%"
```
5. port-forward the gateway-proxy, make some requests against `/` and `/withMetadata` to see the access log differences in the gateway-proxy logs

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by ben-taussig-solo** [comment]
### After building the gloo container image from this branch:

Requests to `/` result in the following log message:
```
gateway-proxy-77966f6fb7-d9gs8 route metadata: "-"; route metadata drilldown: "-"
```
Requests to `/withMetadata` result in the following log message:
```
gateway-proxy-77966f6fb7-d9gs8 route metadata: "{"bar-metadata":"I'm on the gloo route!"}"; route metadata drilldown: "I'm on the gloo route!"
```

File: test/e2e/access_log_test.go [hunk 1]
```diff
@@ -123,19 +138,40 @@ var _ = Describe("Access Log", func() {
...
 					logs, err := testContext.EnvoyInstance().Logs()
 					g.Expect(err).NotTo(HaveOccurred())
-					g.Expect(logs).To(ContainSubstring(`"POST /1 HTTP/1.1" 200`))
+					g.Expect(logs).To(ContainSubstring(`"POST /1?foo=bar HTTP/1.1" 200`))
 				}, time.Second*30, time.Second/2).Should(Succeed())
 			})
+			Context("Formatter extensions", func() {
+				BeforeEach(func() {
+					gw.GetOptions().GetAccessLoggingService().GetAccessLog()[0].GetFileSink().OutputFormat = &als.FileSink_StringFormat{
+						StringFormat: "%REQ(:PATH)%; %REQ_WITHOUT_QUERY(:PATH)%; %METADATA(ROUTE:foo-namespace)%\n",
```
Comment (ben-taussig-solo) [inline test/e2e/access_log_test.go:157]: Nit: Suggestion for improved readability in this test
```suggestion
						StringFormat: "req: %REQ(:PATH)%\n" +
							"req_without_query: %REQ_WITHOUT_QUERY(:PATH)%\n" +
							"metadata: %METADATA(ROUTE:foo-namespace)%\n",
```

File: test/e2e/access_log_test.go [hunk 2]
```diff
@@ -123,19 +138,40 @@ var _ = Describe("Access Log", func() {
...
+					requestBuilder := testContext.GetHttpRequestBuilder().
+						WithPath("1").
+						WithQuery("sensitive=data&needs=removed").
+						WithPostMethod()
+					Eventually(func(g Gomega) {
+						g.Expect(testutils.DefaultHttpClient.Do(requestBuilder.Build())).Should(matchers.HaveOkResponse())
+
+						logs, err := testContext.EnvoyInstance().Logs()
+						g.Expect(err).NotTo(HaveOccurred())
+						g.Expect(logs).To(ContainSubstring(`/1?sensitive=data&needs=removed; /1; {"bar-metadata":"greetings"}`))
```
Comment (ben-taussig-solo) [inline test/e2e/access_log_test.go:170]: nit: second part of [above suggestion](https://github.com/solo-io/gloo/pull/9247/files#r1532149643)
```suggestion
						g.Expect(logs).To(ContainSubstring(`req: /1?sensitive=data&needs=removed`))
						g.Expect(logs).To(ContainSubstring(`req_without_query: /1`))
						g.Expect(logs).To(ContainSubstring(`metadata: {"bar-metadata":"greetings"}`))
```

**Comment by ben-taussig-solo** [comment]
Is there a deprecation timeline for the other directives (`%DYNAMIC_METADATA()%`, etc.)?

**Comment by jbohanon** [comment]
> Is there a deprecation timeline for the other directives (`%DYNAMIC_METADATA()%`, etc.)?

That's a good question. It looks like the person that did the original implementation just added the "deprecated in the future" comment https://github.com/envoyproxy/envoy/pull/17457.

I will ask if there is a method to move that along

**Comment by jbohanon** [comment]
Opened https://github.com/envoyproxy/envoy/issues/33004

**Review by ben-taussig-solo — APPROVED** [review]
> > Is there a deprecation timeline for the other directives (`%DYNAMIC_METADATA()%`, etc.)?
> 
> That's a good question. It looks like the person that did the original implementation just added the "deprecated in the future" comment [envoyproxy/envoy#17457](https://github.com/envoyproxy/envoy/pull/17457).
> 
> I will ask if there is a method to move that along. Also since the extension is like >2 years old it might be able to come off of WIP status and be marked production ready

Gotcha, makes sense. I was looking through some upstream envoy GH issues and it didn't seem like there was much recent discussion about this, so I don't think we need to be too concerned about this stuff shifting under us

Anyway this PR looks good to me -- I just approved but ping me if you end up making any changes and need another review

File: test/e2e/access_log_test.go [hunk 3]
```diff
@@ -46,6 +46,21 @@ var _ = Describe("Access Log", func() {
 
 	JustBeforeEach(func() {
 		testContext.JustBeforeEach()
+		testContext.PatchDefaultVirtualService(func(vs *v1.VirtualService) *v1.VirtualService {
```
Comment (sam-heilbron) [inline test/e2e/access_log_test.go:49]: nit: The way this is currently implemented, we write a default vs, and then read it, modify it and write the updated value.

An alternative, would be in the BeforeEach, to do something like https://github.com/solo-io/gloo/blob/main/test/e2e/cors_test.go#L73, where we either define the exact VS we need, or just modify the default one that is supplied

File: changelog/v1.17.0-beta13/metadata-access-log-formatter.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+changelog:
+- type: NEW_FEATURE
+  issueLink: https://github.com/solo-io/gloo/issues/9246
+  resolvesIssue: false
+  description: >-
+    Add ability to use `%METADATA()%` command directives in envoy access log format
```
Comment (sam-heilbron) [inline changelog/v1.17.0-beta13/metadata-access-log-formatter.yaml:6]: are there any updates to our user guides that was could/should make to demonstrate this?

**Review by sam-heilbron — DISMISSED** [review]
LGTM

File: changelog/v1.17.0-beta13/metadata-access-log-formatter.yaml (hunk 4 above)
Comment (jbohanon) [reply changelog/v1.17.0-beta13/metadata-access-log-formatter.yaml:6]: I thought about this, but I don't think we need to. A user would need to be familiar with Envoy access logging commands in order to be configuring their own log anyway, so I think we are just net-positive in the sense that if they read about this, want to use it and try to it "just works" without them having to be concerned about it coming from an extension or anything"

**Review by sam-heilbron — APPROVED** [review]

**Review by ben-taussig-solo — APPROVED** [review]

**Review by ben-taussig-solo — APPROVED** [review]

**Review by sam-heilbron — APPROVED** [review]

**Pull request merged**